src/idpi/product.py

```diff
@@ -0,0 +1,58 @@
+"""Product base classes."""
+
+# Standard library
+import typing
+from abc import ABCMeta
+from abc import abstractmethod
+from itertools import accumulate
+
+# Third-party
+import dask
+
+
+class Register:
+    """Register methods that are (dask) delayed for caching."""
+
+    def __init__(self, delay: bool = False):
+        self.regdict: dict[str, typing.Any] = {}
+        self._delayed = dask.delayed if delay else lambda x: x
+
+    def reg(self, fn, *arg):
+        key = list(
+            accumulate([hash(id(x)) for x in arg], lambda x, acc: hash(x ^ acc))
+        )[-1]
```

Comment:
Consider using a frozenset of the ids and hashing that. Note that the pattern `list(accumulate(...))[-1]` is equivalent to `reduce(...)`.

Reply:
great suggestion thanks